cli/src/cli/keygen.rs

```diff
@@ -1,46 +1,105 @@
-use error::DefaultResult;
 use holochain_common::paths::keys_directory;
-use holochain_conductor_api::{key_loaders::mock_passphrase_manager, keystore::Keystore};
-use rpassword;
+use holochain_conductor_api::{
+    key_loaders::mock_passphrase_manager,
+    keystore::{Keystore, PRIMARY_KEYBUNDLE_ID},
+};
+use holochain_core_types::error::HcResult;
+use holochain_dpki::seed::{SeedType, TypedSeed};
 use std::{
     fs::create_dir_all,
     io::{self, Write},
     path::PathBuf,
+    sync::{Arc, Mutex},
 };
+use util::{get_secure_string_double_check, get_seed, user_prompt};
 
-pub fn keygen(path: Option<PathBuf>, passphrase: Option<String>, quiet: bool) -> DefaultResult<()> {
-    let passphrase = passphrase.unwrap_or_else(|| {
-        if !quiet {
-            println!(
-                "
-This will create a new agent keystore and populate it with an agent keybundle
+pub fn keygen(
+    path: Option<PathBuf>,
+    keystore_passphrase: Option<String>,
+    nullpass: bool,
+    mnemonic_passphrase: Option<String>,
+    root_seed_mnemonic: Option<String>,
+    device_derivation_index: Option<u64>,
+    quiet: bool,
+) -> HcResult<()> {
+    user_prompt(
+        "This will create a new agent keystore and populate it with an agent keybundle
 containing a public and a private key, for signing and encryption by the agent.
 This keybundle will be stored encrypted by passphrase within the keystore file.
 The passphrase is securing the keys and will be needed, together with the file,
-in order to use the key.
-Please enter a secret passphrase below. You will have to enter it again
-when unlocking the keybundle to use within a Holochain conductor."
+in order to use the key.\n",
+        quiet,
+    );
+
+    let keystore_passphrase = match (keystore_passphrase, nullpass) {
+        (None, true) => String::from(holochain_common::DEFAULT_PASSPHRASE),
+        (Some(s), false) => s,
+        (Some(_), true) => panic!(
+            "Invalid combination of args. Cannot pass --nullpass and also provide a passphrase"
+        ),
+        (None, false) => {
+            // prompt for the passphrase
+            user_prompt(
+                "Please enter a secret passphrase below. You will have to enter it again
+when unlocking the keybundle to use within a Holochain conductor.\n",
+                quiet,
```

Comment:
What do you think of utilizing combinators for this one and just utilizing the null_pass boolean?

keystore_passphrase
.map(|s|{
    if nullpass = true{panic!("do this)}else{s}
})
.unwrap_or_else(||{
     //userprompt
	 if nullpass = true
     {
	     get_secure_string_double_check("keystore Passphrase", quiet)
	     .expect("Could not retrieve passphrase")
     }
     else
     {
         String::from(holochain_common::DEFAULT_PASSPHRASE)
	 }	 
	 
})